libc/include/llvm-libc-types/float128.h

```diff
@@ -0,0 +1,45 @@
+//===-- Definition of float128 type ---------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_TYPES_FLOAT128_H__
+#define __LLVM_LIBC_TYPES_FLOAT128_H__
+
+#include <include/llvm-libc-macros/float-macros.h> // LDBL_MANT_DIG
+
+// Define temporary compiler and its version
+#if defined(__clang__)
+#define __LIBC_COMPILER_IS_CLANG
+#define __LIBC_COMPILER_CLANG_VER (__clang_major__ * 100 + __clang_minor__)
+#elif defined(__GNUC__)
+#define __LIBC_COMPILER_IS_GCC
+#define __LIBC_COMPILER_GCC_VER (__GNUC__ * 100 + __GNUC_MINOR__)
+#endif // __clang__, __GNUC__
+
+#if (defined(__LIBC_COMPILER_GCC_VER) && (__LIBC_COMPILER_GCC_VER >= 1301)) && \
+    (defined(__aarch64__) || defined(__riscv) || defined(__x86_64__))
+#define LIBC_COMPILER_HAS_C23_FLOAT128
+typedef _Float128 float128;
+#elif (defined(__LIBC_COMPILER_CLANG_VER) &&                                   \
+       (__LIBC_COMPILER_CLANG_VER >= 600)) &&                                  \
+    (defined(__x86_64__) && !defined(__Fuchsia__))
+#define LIBC_COMPILER_HAS_FLOAT128_EXTENSION
+typedef __float128 float128;
```

Comment:
Is there a bug filed against clang upstream that you could link to in a comment here regarding clang support for _Float128?

For instance, I suspect `_Float128` is what was standardized, not `__float128`.  It would be good to have a TODO in sources to fix this up in the future, or think about how to write it such that newer versions of clang that do support `_Float128` just work.

Reply:
Unfortunately, the current landscape of C23 _Float128 type is quite messy.  I added a TODO for us to update this detection once clang has `_Float128` type on our platforms of interest.

Reply:
Unfortunately, the current landscape of C23 _Float128 type is quite messy.  I added a TODO for us to update this detection once clang has `_Float128` type on our platforms of interest.

Reply:
To my knowledge there is no principled way to test if a compiler specific type is supported.
From the standard point of view we need C++23 but [LLVM libc is still C++17](https://github.com/llvm/llvm-project/blob/73e546625d6567e5a33454741d257b6929c89635/libc/CMakeLists.txt#L30)
https://en.cppreference.com/w/cpp/types/floating-point

Reply:
Please add a link to https://github.com/llvm/llvm-project/issues/80195.

`__STDC_IEC_60559_BFP__` and `__STDC_IEC_60559_TYPES__` are what we should use for feature detection.

Reply:
Please add a link to https://github.com/llvm/llvm-project/issues/80195 though.

Reply:
I added the link to it right next to the `TODO`.